tests/test_metrics.py

```diff
@@ -0,0 +1,174 @@
+        [-0.75616974, -0.6701199 , -0.5893732 , -0.5242175 ,
+         -0.4719131 , -0.42543185, -0.43560237, -0.45590907,
+         -0.51438874, -0.61130494, -0.74402857]],
+       [[-0.18297303, -0.21213517, -0.18341143, -0.15066521,
+         -0.11950047, -0.09036797, -0.0256229 ,  0.0269562 ,
+          0.06986493,  0.1414077 ,  0.19563401],
+        [-0.18315202, -0.21226275, -0.18336335, -0.15038337,
+         -0.11897573, -0.08946133, -0.0242492 ,  0.02882081,
+          0.07219976,  0.14433557,  0.19909364]],
+       [[ 0.36912787,  0.34506714,  0.32219756,  0.3202601 ,
+          0.30032292,  0.259299  ,  0.21430482,  0.14271711,
+          0.05134173, -0.063667  , -0.17867568],
+        [ 0.36715215,  0.34335977,  0.32078195,  0.3192455 ,
+          0.2996201 ,  0.2587561 ,  0.21395013,  0.14270164,
+          0.05165949, -0.06302758, -0.1777146 ]]])
+
+# An example of scores obtained for a swarm that bounce around on the way down.
+epoch_scores = [0.51727545, 0.4584964 , 0.3589881 , 0.2524824 , 0.20734829,
+       0.2482427 , 0.30246153, 0.3388226 , 0.34041768, 0.3064342 ,
+       0.26800793, 0.2686419 , 0.24010916, 0.18522426, 0.22644123,
+       0.26727045, 0.28942722, 0.28332102, 0.25410518, 0.22259913,
+       0.25512502, 0.28029743, 0.29604492, 0.30136263, 0.29408443,
+       0.27543014, 0.24885914, 0.21919054, 0.22593765, 0.2305434 ,
+       0.22474495, 0.21082267, 0.19170743, 0.17090012, 0.1521816 ,
+       0.13839552, 0.1299243 , 0.12569669, 0.12456866, 0.12922356,
+       0.14023647, 0.15060309, 0.15662336, 0.15730526, 0.15512368,
+       0.15510257, 0.16903949, 0.1815229 , 0.20310307, 0.21428823,
+       0.21110815, 0.19391632, 0.16897929, 0.15510854, 0.1513776 ,
+       0.15778454, 0.15062831, 0.1423014 , 0.1533089 , 0.16309854]
+
+
+def test_summarise_across_bees_ypreds():
+    for summ_metric in ['min', 'max', 'mean', 'median', 'std', 'range']:
+        out = metrics.summarise_across(ypreds, summ_metric)
+        assert type(out) == np.ndarray
+        assert out.shape == (2, 11)
+
+
+def test_summarise_across_bees_losses():
+    for summ_metric in ['min', 'max', 'mean', 'median', 'std', 'range']:
+        out = metrics.summarise_across(losses, summ_metric)
+        assert type(out) == np.ndarray
+        assert out.shape == (2,)
+
+
+def test_rmse_2d():
+    b0_preds = ypreds[0]
+    out = metrics.rms_error(b0_preds, y)
+    assert len(out.shape) == len(b0_preds.shape) - 1
+    assert np.max(np.abs(out - losses[0])) < 0.00001  # I don't know why these aren't exactly 0
+
+    b2_preds = ypreds[2]
+    out = metrics.rms_error(b2_preds, y)
+    assert len(out.shape) == len(b2_preds.shape) - 1
+    assert np.max(np.abs(out - losses[2])) < 0.00001  # I dont' know why this isn't exactly 0
+
+
+def test_rmse_3d():
+    out = metrics.rms_error(ypreds, y)
+    assert len(out.shape) == len(ypreds.shape) - 1
+    assert np.max(np.abs(out - losses)) < 0.00001  # I don't know why this isn't exactly 0
```

Comment:
floating point error. use `pytest.approx(float1) == float2`

Reply:
Excellent.

Reply:
https://www.youtube.com/watch?v=PZRI1IfStY0 a great video on the topic

Reply:
This didn't actually work. I tried it.  But `pytest.approx` still applies too tight a test. 
```
def test_rmse_2d():
        b0_preds = ypreds[0]
        out = metrics.mse_loss(b0_preds, y)
        assert len(out.shape) == len(b0_preds.shape) - 1
>       assert approx(np.max(np.abs(out - losses[0]))) == 0
E       assert 6.601767454927909e-08 ± 1.0e-12 == 0
```

Reply:
It can be annoying to set right - especially around 0 since you don't get relative error. I suspect pytest approx is doing it for float64 when numpy is using float32 by default - https://docs.pytest.org/en/stable/reference.html#pytest-approx 
Try doing `pytest.approx(..., abs = 1e-7)`